File: apps/ui/lib/ui_web/graph/mutations/event.ex
defmodule UIWeb.Graph.Schema.Mutations.Event do
  use Absinthe.Schema.Notation

  alias Accounts.Authentication.InvalidPermission
  alias Events.Event
  alias UIWeb.Graph.Schema.Utils.InvalidParams
  alias UIWeb.Graph.Schema.Resolvers.Event, as: EventResolver

  input_object :create_event_input do
    field :name, non_null(:string)
    field :event_start, non_null(:datetime)
    field :event_end, non_null(:datetime)
    field :sheet_open, non_null(:datetime)
    field :sheet_closed, non_null(:datetime)
    field :sheet_cost, :integer
    field :funds_granted, :integer
  end

  union :create_event_result do
    types [:event, :invalid_params, :invalid_permissions]

    resolve_type fn
      %Event{}, _ -> :event
      %InvalidPermission{}, _ -> :invalid_permissions
      %InvalidParams{}, _ -> :invalid_params
    end
  end

  object :event_mutations do
    field :create_event, non_null(:create_event_result) do
      arg :input, non_null(:create_event_input)
      resolve &EventResolver.create_event/2
    end

    field :update_event, non_null(:event) do
      arg :id, non_null(:id)
      arg :name, :string
      arg :event_start, :datetime
      arg :event_end, :datetime
      arg :sheet_open, :datetime
      arg :sheet_closed, :datetime
      arg :sheet_cost, :integer
      arg :funds_granted, :integer
      resolve &EventResolver.update/2
    end

    field :add_athlete_to_event, non_null(:event) do
      arg :event_id, non_null(:id)
      arg :athlete_id, non_null(:id)
      arg :cost, non_null(:float)
      resolve &EventResolver.add_athlete/2
    end

    field :update_event_athlete, non_null(:event_athlete) do
      @desc "The id for the event athlete"
      arg :id, non_null(:id)
      arg :cost, non_null(:float)
      arg :winnings, non_null(:float)
      resolve &EventResolver.update_event_athlete/2
    end

    field :remove_event_athlete, non_null(:event) do
      @desc "Id for the event athlete to remove"
      arg :id, non_null(:id)

      resolve &EventResolver.remove_event_athlete/2
    end
  end
end

File: apps/ui/assets/js/stores/event.js
import { writable } from 'svelte/store'
import api from '@api'
import Event from '../models/events/event'

const EVENT_QUERY = `
  query eventById($id: ID!) {
    event(id: $id) {
      id
      name
      eventStart
      eventEnd
      sheetOpen
      sheetClosed
      sheetCost
      fundsGranted
    }
  }
`

const UPDATE_EVENT_DETAILS_QUERY = `
    mutation updateEvent($id: ID!, $name: String, $eventStart: DateTime, $eventEnd: DateTime, $sheetOpen: DateTime, $sheetClosed: DateTime, $sheetCost: Int, $fundsGranted: Int) {
      updateEvent(id: $id, name: $name, eventStart: $eventStart, eventEnd: $eventEnd, sheetOpen: $sheetOpen, sheetClosed: $sheetClosed, sheetCost: $sheetCost, fundsGranted: $fundsGranted) {
          id
          name
          eventStart
          eventEnd
          sheetOpen
          sheetClosed
          sheetCost
          fundsGranted
        }
    }
`

const createEventStore = () => {
  const { subscribe, update } = writable({loading: false, event: null})

  return {
    subscribe,
    getById(id) {
      update(store => ({...store, loading: true}))
      return api.graph(EVENT_QUERY, { id })
        .then(({ event }) =>
          update(store => ({...store, event: new Event(event)}))
        )
        .finally(() => update(store => ({...store, loading: false})))
    },
    updateDetails(event) {
      update(store => ({...store, loading: true}))

      return api.graph(UPDATE_EVENT_DETAILS_QUERY, {id: event.id, ...event.toJSON()})
        .then(({ updateEvent }) => {
          update(store => ({...store, event: new Event(updateEvent)}))
        })
        .finally(() => {
          return update(store => ({...store, loading: false}))
        })
        
    }
  }
}

const eventStore = createEventStore()

export default eventStore
